### PR TITLE
fix: ordering of functionality for setting and evaluating label expressions

### LINKED ISSUE #9625: sprig expression that works in 3.3.9 doesn't work in 3.4.0
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

a param to my workflow is a payload like the json in these 12 lines:
https://github.com/LeidosLabs/holeshot/blob/b3aa04a84059e3a4c07d42ff9a46cc37f443eb3d/ingest-service/src/test/resources/sample-notification.json#L1-L12

my expression below worked fine in 3.3.9 but in 3.4.0 gives error

```
                workflowMetadata:
                  labelsFrom:
                    s3-folder:
                      expression: "{{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}"
```

```
Error: failed to evaluate label "s3-folder" expression "{{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}": a map key must be a quoted string, a number, a identifier, or an expression enclosed in parentheses (unexpected token Bracket("{")) (1:2) | {{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }} | .^
```

### Version

3.4.0

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
n/a
```

### Logs from the workflow controller

kubectl logs -n argo deploy/workflow-controller | grep ${workflow}

```
time="2022-09-20T00:57:22.742Z" level=info msg="Processing workflow" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Updated phase  -> Error" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Updated message  -> failed to evaluate label \"s3-folder\" expression \"{{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}\": a map key must be a quoted string, a number, a identifier, or an expression enclosed in parentheses (unexpected token Bracket(\"{\")) (1:2)\n | {{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}\n | .^" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Marking workflow completed" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Marking workflow as pending archiving" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=error msg="Unable to set ExecWorkflow" error="failed to evaluate label \"s3-folder\" expression \"{{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}\": a map key must be a quoted string, a number, a identifier, or an expression enclosed in parentheses (unexpected token Bracket(\"{\")) (1:2)\n | {{= sprig.quote(sprig.regexReplaceAll('^[-_.]*',sprig.regexReplaceAll('[-_.]*$',sprig.substr(0,63,sprig.regexReplaceAll('[^-A-Za-z0-9_.]',sprig.regexReplaceAll('/([^/]*)$',sprig.regexFind('landing/([^,]+),', workflow.parameters.message),''),'')),''),'')) }}\n | .^" namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Checking daemoned children of " namespace=auth workflow=redacted
time="2022-09-20T00:57:22.749Z" level=info msg="Workflow to be dehydrated" Workflow Size=44708

```

### Logs from in your workflow's wait container

kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

### PR DESCRIPTION
**Pull request by juliev0**
Fixes #9625 

Fixing the order of functionality for setting and evaluating label expressions, to address the issue that a _LabelsFrom_ expression may be dependent on the evaluation of a parameter that it uses; as well as the issue that workflow.labels.x may be used somewhere else in the spec. The new order is:

1. Set _globalParams_ and substitute them in the spec **for everything except _WorkflowMetadata.LabelsFrom_ expressions** (which will be done in step 3)
2. Evaluate _WorkflowMetadata.LabelsFrom_ expressions and set _globalParams_ for those
3. Perform any substitution in the spec based on these new _globalParams_

### CODE REVIEW

File: workflow/controller/testdata/workflow-sub-test-6.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  synchronization:
+    mutex:
+      name:  "{{workflow.labels.mutexName}}"
+  schedulerName:  "{{workflow.annotations.schedulerName}}"
+  workflowMetadata:
+    labelsFrom:
+      mutexName: 
+        expression: "{{= sprig.quote(sprig.trim( workflow.annotations.schedulerName )) }}"
+    annotations:
+      schedulerName: wfMetadataScheduler
```
Comment (juliev0) [inline workflow/controller/testdata/workflow-sub-test-6.yaml:22]: In this new test, we verify that:

1. workflowMetadata.labelsFrom.mutexName.expression (which is based on another parameter) gets evaluated correctly
2. the result of that label gets incorporated into spec.synchronization.mutex.name

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -570,6 +586,23 @@ func (woc *wfOperationCtx) setGlobalParameters(executionParameters wfv1.Argument
...
 		}
 	}
+	if woc.wf.Status.Outputs != nil {
+		for _, param := range woc.wf.Status.Outputs.Parameters {
+			if param.HasValue() {
+				woc.globalParams["workflow.outputs.parameters."+param.Name] = param.GetValue()
+			}
+		}
+	}
+
```
Comment (juliev0) [inline workflow/controller/operator.go:596]: note this isn't new code - just relocated

**Review by tooptoop4 — APPROVED** [review]

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -482,22 +482,30 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
...
 		}
-		if woc.wf.ObjectMeta.Annotations == nil {
-			woc.wf.ObjectMeta.Annotations = make(map[string]string)
+		if woc.wf.Annotations == nil {
+			woc.wf.Annotations = make(map[string]string)
 		}
 		for n, v := range md.Annotations {
 			woc.wf.Annotations[n] = v
-			woc.globalParams["workflow.annotations."+n] = v
+			woc.globalParams["workflow.labels."+n] = v
```
Comment (sarabala1979) [inline workflow/controller/operator.go:505]: wrong key `workflow.annotations`

Comment (juliev0) [reply workflow/controller/operator.go:505]: oops, good catch

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**